DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/state_machine/src/state_machine.hpp

```diff
@@ -0,0 +1,133 @@
+#ifndef STATE_MACHINE_HPP
+#define STATE_MACHINE_HPP
+
+#include <sycl/ext/intel/experimental/task_sequence.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/pipes_ext.hpp>
+#include <sycl/sycl.hpp>
+
+enum State {
+  PROCESS,
+  LD_COEFF
+};
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class IDNaiveStateMachine;
+class IDOptimizedStateMachine;
+class IDStreamDataIn_NaiveStateMachine;
+class IDStreamDataIn_OptimizedStateMachine;
+class IDStreamDataOut_NaiveStateMachine;
+class IDStreamDataOut_OptimizedStateMachine;
+
+// define data pipes
+using StreamingPipeProps = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::uses_valid<true>,
+    sycl::ext::intel::experimental::bits_per_symbol<32>,
+    sycl::ext::intel::experimental::protocol<
+        sycl::ext::intel::experimental::protocol_name::
+            avalon_streaming_uses_ready>));
+
+using MyStreamingBeat =
+    sycl::ext::intel::experimental::StreamingBeat<float, true, false>;
+
+using StreamIn_NaiveStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataIn_NaiveStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>;
+using StreamIn_OptimizedStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataIn_OptimizedStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>; 
+using StreamOut_NaiveStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataOut_NaiveStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>;
+using StreamOut_OptimizedStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataOut_OptimizedStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>;
+
+// function for task_sequence
+float UpdateCoeff(float *coeff_ptr, int *execution_count) {
+  float coeff_new = coeff_ptr[0];
+  // read-modify-write of external memory will result in a poor II.
+  execution_count[0] = execution_count[0] + 1;
+  return coeff_new;
+}
+
+// function for task_sequence
+float Compute(float coeff, float data) {
+  return coeff * data;
+}
+
+struct NaiveStateMachine {
+  float *coeff_ptr;
+  int *execution_count;
+  bool init_coeff_before_starting = false;
+  int iterations;
+
+  void operator()() const {
+    State my_state = 
+        (init_coeff_before_starting) ? State::LD_COEFF : State::PROCESS;
+    float coeff = 1.0f;
+
```

Comment:
In the optimized design, this loop has `[[intel::initiation_interval(1)]]`, I think we can add a comment here explain this loop cannot achieve II=1.